libraries/botbuilder/tests/teams/composeMessagingExtensionAuth/src/composeMessagingExtensionAuthBot.ts

```diff
@@ -0,0 +1,209 @@
+*/
+export class ComposeMessagingExtensionAuthBot extends TeamsActivityHandler {
+    connectionName: string;
+    constructor(authConnectionName: string) {
+        super();
+        
+        this.connectionName = authConnectionName;
+
+        // See https://aka.ms/about-bot-activity-message to learn more about the message and other activity types.
+        this.onMessage(async (context, next) => {
+
+            // Hack around weird behavior of RemoveRecipientMention (it alters the activity.Text)
+            const originalText = context.activity.text;
+            TurnContext.removeRecipientMention(context.activity);
+            const text = context.activity.text.replace(' ', '').toUpperCase();
+            context.activity.text = originalText;
+
+            if (text === 'LOGOUT' || text === 'SIGNOUT')
+            {
+                const adapter: IUserTokenProvider = context.adapter as BotFrameworkAdapter;
+
+                await adapter.signOutUser(context, this.connectionName);
+                await context.sendActivity(`Signed Out: ${context.activity.from.name}`);
+                
+                return;
+            }
+
+            await context.sendActivity(`You said '${context.activity.text}'`);
+            // By calling next() you ensure that the next BotHandler is run.
+            await next();
+        });
+
+        this.onMembersAdded(async (context, next) => {
+            const membersAdded = context.activity.membersAdded;
+            for (const member of membersAdded) {
+                if (member.id !== context.activity.recipient.id) {
+                    await context.sendActivity('Hello and welcome!');
+                }
+            }
+
+            // By calling next() you ensure that the next BotHandler is run.
+            await next();
+        });
+    }
+
+    protected async onTeamsMessagingExtensionFetchTask(context: TurnContext, query: MessagingExtensionQuery): Promise<MessagingExtensionActionResponse> {
+        const adapter: IUserTokenProvider = context.adapter as BotFrameworkAdapter;
+        const userToken = await adapter.getUserToken(context, this.connectionName);
+        if (!userToken)
+        {
+            // There is no token, so the user has not signed in yet.
+
+            // Retrieve the OAuth Sign in Link to use in the MessagingExtensionResult Suggested Actions
+            const signInLink = await adapter.getSignInLink(context, this.connectionName);
+
+            const response : MessagingExtensionActionResponse = { 
+                composeExtension: { 
+                    type: 'auth', 
+                    suggestedActions: { 
+                        actions: [{ 
+                            type: 'openUrl', 
+                            value: signInLink, 
+                            title: 'Bot Service OAuth' 
+                        }]
+                    }
+                }
+            };
```

Comment:
I got into the practice of casting (so you get some sense of what each block is) - for example:
 public static toComposeExtensionResultResponse(cardAttachment: Attachment) : MessagingExtensionActionResponse {
        
        return <MessagingExtensionActionResponse> {
            composeExtension: <MessagingExtensionResult> {
                type: 'result',
                attachmentLayout: 'list',
                attachments: [ <MessagingExtensionAttachment> cardAttachment ]

            }
        }
    }

Reply:
That would be better ... i've just been casting the top level return, but should be casting the fields too, you're right!